app/models/highlight.rb

```diff
@@ -16,54 +21,113 @@ class Highlight < ApplicationRecord
                         :highlighted_content,
                         :location_strategies
 
-  validate :source_ids_uuid_for_openstax_page
+  validate :must_have_source_type_specific_ids
+  validate :must_have_consistent_neighbors, on: :create
 
   enum source_type: [:openstax_page]
 
-  before_validation :normalize_color, :normalize_source_ids
-
   scope :by_source_type, ->(source_type) { where(source_type: source_type) }
   scope :by_color, ->(color) { where(color: color) }
   scope :by_user, ->(user_id) { where(user_uuid: user_id) }
-  scope :by_order, ->(order_direction) {
-    order_direction = order_direction.to_sym
-    order(order_in_source: order_direction, source_order: order_direction)
-  }
-  scope :by_source_parent_ids, ->(source_parent_ids) do
-    where('source_parent_ids && ?', postgres_style_array(source_parent_ids))
-  end
+  scope :by_scope_id, ->(scope_id) { where(scope_id: scope_id) }
+  scope :by_source_ids, ->(source_ids) { where(source_id: source_ids) }
+
+  around_create :insert_new_highlight_between_neighbors
+  before_destroy :reconnect_neighbors_around_destroyed_highlight
 
   def normalize_color
     self.color = color&.downcase
   end
 
-  def normalize_source_ids
+  def normalize_ids
     if openstax_page?
       self.source_id = source_id&.downcase&.strip
-      source_parent_ids.map!(&:downcase).map!(&:strip)
+      self.scope_id = scope_id&.downcase&.strip
     end
   end
 
-  private
+  protected
+
+  def all_from_scope_and_source
+    Highlight.where(scope_id: scope_id).where(source_id: source_id)
```

Comment:
perhaps this can be built reusing the scopes created above?  or to create a by_scope_source_id scope that uses both?  (maybe the chained version might be reused sometime)

Reply:
I'm ok with this because it is pretty simple and hidden internally.  I do like the `by_blah` scopes as a public mechanism that helps with querying generically.